ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -0,0 +1,335 @@
+    assertTrue("Valid data size collection time should be in the range",
+        timeRange.getStartTimeInMs() <= actualLogSegmentValidSizeMap.getFirst()
+            && timeRange.getEndTimeInMs() >= actualLogSegmentValidSizeMap.getFirst());
+
+    int expectedNumberOfLogSegments = 0;
+    long expectedTotalLogSegmentValidSize = 0L;
+    LogSegment logSegment = state.log.getFirstSegment();
+
+    while (logSegment != null) {
+      String logSegmentName = logSegment.getName();
+      assertTrue("Log segment: " + logSegmentName + " not found",
+          actualLogSegmentValidSizeMap.getSecond().containsKey(logSegmentName));
+
+      long expectedLogSegmentValidSize =
+          state.getValidDataSizeForLogSegment(logSegment, timeRange.getEndTimeInMs(), state.time.milliseconds());
+      long actualLogSegmentValidSize = actualLogSegmentValidSizeMap.getSecond().get(logSegmentName);
+      assertEquals("Valid data size mismatch for log segment: " + logSegmentName, expectedLogSegmentValidSize,
+          actualLogSegmentValidSize);
+
+      expectedTotalLogSegmentValidSize += expectedLogSegmentValidSize;
+      expectedNumberOfLogSegments++;
+      logSegment = state.log.getNextSegment(logSegment);
+    }
+
+    assertEquals("Mismatch in number of log segments", expectedNumberOfLogSegments,
+        actualLogSegmentValidSizeMap.getSecond().size());
+
+    Pair<Long, Long> actualTotalValidSize = blobStoreStats.getValidSize(timeRange);
+    assertTrue("Valid data size collection time should be in the range",
+        timeRange.getStartTimeInMs() <= actualTotalValidSize.getFirst()
+            && timeRange.getEndTimeInMs() >= actualLogSegmentValidSizeMap.getFirst());
+    assertEquals("Total valid data size of all log segments mismatch", expectedTotalLogSegmentValidSize,
+        actualTotalValidSize.getSecond().longValue());
+
+    return actualTotalValidSize.getSecond();
+  }
+
+  /**
+   * Go over the referenceIndex to collect valid data size information per container. The result is used for
+   * verification purposes.
+   * @param deleteAndExpirationRefTimeInMs the reference time in ms until which deletes and expiration are relevant
+   * @return a nested {@link Map} of serviceId to containerId to valid data size
+   */
+  private Map<String, Map<String, Long>> getValidSizeByContainer(long deleteAndExpirationRefTimeInMs) {
+    Map<String, Map<String, Long>> containerValidSizeMap = new HashMap<>();
+    for (Map.Entry<Offset, TreeMap<MockId, IndexValue>> segmentEntry : state.referenceIndex.entrySet()) {
+      for (Map.Entry<MockId, IndexValue> indexEntry : segmentEntry.getValue().entrySet()) {
+        MockId key = indexEntry.getKey();
+        IndexValue value = indexEntry.getValue();
+        if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+          if (!state.isDeletedAt(key, deleteAndExpirationRefTimeInMs) && !state.isExpiredAt(key, deleteAndExpirationRefTimeInMs)
```

Comment:
why not use getValidIndexEntriesForIndexSegment from CuratedLogIndexState instead of again trying to find valid entries?